storage/registers.go

```diff
@@ -0,0 +1,33 @@
+package storage
+
+import (
+	"io"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// Registers defines methods for the register index.
+type Registers interface {
+	RegisterReader
+	RegisterWriter
+	io.Closer
+}
+
+// RegisterReader defines read-only operations on the register index.
+type RegisterReader interface {
+	// Get register by the register ID at a given block height.
+	//
+	// If the register at the given height was not indexed, returns the highest
+	// height the register was indexed at.
+	// An error is returned if the register was not indexed at all. Expected errors:
+	// - storage.ErrNotFound if the register was not found in the db
```

Comment:
The storage needs to keep track of the latest height, and the range is `[root_block_height, latest_height]`.

The latest_height can be initialized as `root_block_height`, meaning on startup, we check if pebble has the `latest_height` key, if no, it means the pebble db needs to be bootstrapped, and we set `root_block_height` as `latest_height`. 

When we save registers by height, we save update the `latest_height` along with the registers. 

On startup, we also read the latest_height into memory, and keep track it using a variable. The cached `latest_height` will be updated after storing registers into pebble. 

The cached `latest_height` can be used to determine the range. `Get` method will first check if the height is beyond the range, if yes, we can return NotFound error.

Reply:
Makes sense. The only follow-up question I have is whether this is an implementation detail leaking into the registers db? what I mean by this is that currently the indexer uses the jobqueue which persists progress and that is the latest height we can retrieve. If we keep another version of the latest height we processed in Pebble I can see this as a potential mismatch or syncing issue in case of failures, furthermore, it would require duplication of the logic that the jobqueue already has about syncing the progress. But then I can see how having this as part of register db is valuable to prevent lookups that are beyond the highest height and in the context of index db the progress db is not available. Just wanted to raise this.

Reply:
> register not found. The height has been indexed, but the register is not found at the given height

I don't think this will work. We need to return the register even if the requesting height is later than the register was indexed at. This is because if we want to execute scripts at the latest height, all the registers needed by the script will be indexed at previous heights.

Reply:
Yes, there will be three types of errors, which we need to distinguish:
1. storage.HeightNotIndexed, when the height is not indexed
for instance, `Block_10(X:1, Y:2) <- Block_11(X:2)`, then getting register `X` at Block 12 should return `storage.HeightNotIndexed`. Note, we should not return 2, because X might be updated at Block 12.

2. storage.ErrNotFound, when the height has been indexed, and register is not found.
for instance, `Block_10(X:1, Y:2) <- Block_11(X:2)`, then getting register `Z` at Block 11 should return `storage.ErrNotFound`, because Z does not exist not long in block 10 and 11, but also not updated since genesis block.

3. error, any exception

Reply:
I completely know what you mean about different error types. The thing I struggle with understanding how it would work is the case of returning 1st error instead of actually returning registers. 
Let me give an example. I want to execute a script that would read value `X` at the latest block height. The script execution will request a register storing the value `X` at the latest height and the storage layer has to return (in my understanding) the value of `X` even if it was indexed Y heights ago. We will never have all registers indexed at latest height since they don't "carry" over to new heights. So in your case the storage would return an error and the script execution would fail because the register wasn't indexed at that specific height. Not sure how this would work.

Reply:
@sideninja the check would be against the global highest indexed height, not the highest for that specific register. Essentially, the underlying code would check the input height against the last value passed to `SetLatestHeight` and return an error immediately.

After our discussion the other day, I've convinced myself that the register storage does need the be aware of the indexed range and reply with more precise errors for safety. Otherwise, it would be easy for a developer to make the incorrect assumption that not found means the register didn't exist, resulting in successful but incorrect script execution results.